src/v/wasm/transform_module_v2.cc

```diff
@@ -0,0 +1,227 @@
+        max_size = std::max(payload_size, max_size);
+        parser.skip(payload_size);
+        model::timestamp ts = input.header().max_timestamp;
+        if (
+          input.header().attrs.timestamp_type()
+          == model::timestamp_type::create_time) {
+            ts = model::timestamp(
+              input.header().first_timestamp() + timestamp_delta);
+        }
+        records.push_back({
+          .metadata_size = rs_amt + meta_size,
+          .payload_size = payload_size,
+          .attributes = model::record_attributes(attrs),
+          .timestamp = ts,
+          .offset = input.base_offset() + offset_delta,
+        });
+    }
+
+    _call_ctx.emplace(batch_transform_context{
+      .batch_header = input.header(),
+      .batch_data = std::move(input).release_data(),
+      .max_input_record_size = max_size,
+      .records = std::move(records),
+      .output_data = {},
+      .record_callback = std::move(cb),
+    });
+
+    co_await host_wait_for_proccessing();
+
+    auto result = std::exchange(_call_ctx, std::nullopt);
+    co_return std::move(result->output_data);
+}
+
+void transform_module_v2::check_abi_version_1() {
+    // This function does nothing at runtime, it's only an opportunity for
+    // static analysis of the module to determine which ABI version to use.
+}
+
+// NOLINTBEGIN(bugprone-easily-swappable-parameters)
+ss::future<int32_t> transform_module_v2::read_batch_header(
+  int64_t* base_offset,
+  int32_t* record_count,
+  int32_t* partition_leader_epoch,
+  int16_t* attributes,
+  int32_t* last_offset_delta,
+  int64_t* base_timestamp,
+  int64_t* max_timestamp,
+  int64_t* producer_id,
+  int16_t* producer_epoch,
+  int32_t* base_sequence) {
+    // NOLINTEND(bugprone-easily-swappable-parameters)
+
+    co_await guest_wait_for_batch();
+
+    if (!_call_ctx) {
+        co_return NO_ACTIVE_TRANSFORM;
+    }
+    const model::record_batch_header& header = _call_ctx->batch_header;
+    *base_offset = header.base_offset();
+    *record_count = header.record_count;
+    *partition_leader_epoch = int32_t(header.ctx.term());
+    *attributes = header.attrs.value();
+    *last_offset_delta = header.last_offset_delta;
+    *base_timestamp = header.first_timestamp();
+    *max_timestamp = header.max_timestamp();
+    *producer_id = header.producer_id;
+    *producer_epoch = header.producer_epoch;
+    *base_sequence = header.base_sequence;
+
+    _wasi_module->set_timestamp(
+      header.attrs.timestamp_type() == model::timestamp_type::create_time
+        ? header.first_timestamp
+        : header.max_timestamp);
+
+    co_return _call_ctx->max_input_record_size;
+}
+
+ss::future<int32_t> transform_module_v2::read_next_record(
+  uint8_t* attributes,
+  int64_t* timestamp,
+  model::offset* offset,
+  ffi::array<uint8_t> buf) {
+    if (!_call_ctx || _call_ctx->records.empty()) {
+        co_return NO_ACTIVE_TRANSFORM;
+    }
+    co_await ss::coroutine::maybe_yield();
```

Comment:
if this is in a hot path, you may want to consider:

```
if (seastar::need_preempt()) { co_await ss::maybe_yield; }
```

well or not using a coroutine at all. in tight loops i've found the co_wait, even if it immediately returns a ready future can really slow stuff down.

Reply:
Now that we have a microbenchmark I can test :)

Reply:
It seems like in my microbenchmark it's within the noise worth of a change, so I don't think it's worth it, maybe we could write all this code without coroutines? Meh I'll leave it for another day :)